### ISSUE TITLE
Experiments

### ISSUE THREAD

**Issue by OmriDan** [OmriDan/NLP_final_project#6]
# Regression
## Running on small models
small llm, all datasets 
model used = difficulty_regressor_artifacts_mse0.0522.pkl
Mean Squared Error: 0.0737
Mean Absolute Error: 0.2204
R² Score: -0.5426

--- Linear Regression Baseline ---
Creating TF-IDF features...
Training linear regression model...
Linear Regression MSE: 0.0524
Linear Regression MAE: 0.1992
Linear Regression R² Score: -0.2824

--- Comparison to Baseline ---
MSE: -40.54% improvement over baseline
MAE: -10.66% improvement over baseline
R²: -92.17% improvement over baseline

Statistics:
Mean Actual Difficulty: 0.4973
Mean Predicted Difficulty: 0.4468
Std Dev Actual: 0.2201
Std Dev Predicted: 0.1433

model used=difficulty_regressor_artifacts_mse0.0533.pkl
small llms, squad only as dataset
Mean Squared Error: 0.0700

Statistics:
Mean Actual Difficulty: 0.4973
Mean Predicted Difficulty: 0.3594
Std Dev Actual: 0.2201
Std Dev Predicted: 0.0328

model used=difficulty_regressor_artifacts_mse0.0526.pkl
Mean Squared Error: 0.0685

Statistics:
Mean Actual Difficulty: 0.4973
Mean Predicted Difficulty: 0.6156
Std Dev Actual: 0.2201
Std Dev Predicted: 0.0997

# Model with Weighted MSE
model used=difficulty_regressor_artifacts_mse0.0531.pkl
Mean Squared Error: 0.0712

Statistics:
Mean Actual Difficulty: 0.4973
Mean Predicted Difficulty: 0.6156
Std Dev Actual: 0.2201
Std Dev Predicted: 0.1103

## Running on large models (high overfit with training data)
squad only
model used = difficulty_regressor_artifacts_mse0.0541.pkl
Mean Squared Error: 0.0647
Mean Absolute Error: 0.1982
R² Score: -0.3535

--- Linear Regression Baseline ---
Creating TF-IDF features...
Training linear regression model...
Linear Regression MSE: 0.0524
Linear Regression MAE: 0.1992
Linear Regression R² Score: -0.2824

--- Comparison to Baseline ---
MSE: -23.31% improvement over baseline
MAE: 0.49% improvement over baseline
R²: -25.19% improvement over baseline

Statistics:
Mean Actual Difficulty: 0.4973
Mean Predicted Difficulty: 0.3953
Std Dev Actual: 0.2201
Std Dev Predicted: 0.0479

all data
model used = difficulty_regressor_artifacts_mse0.0527.pkl

Mean Squared Error: 0.0746
Mean Absolute Error: 0.2209
R² Score: -0.5622

--- Linear Regression Baseline ---
Creating TF-IDF features...
Training linear regression model...
Linear Regression MSE: 0.0524
Linear Regression MAE: 0.1992
Linear Regression R² Score: -0.2824

--- Comparison to Baseline ---
MSE: -42.33% improvement over baseline
MAE: -10.93% improvement over baseline
R²: -99.12% improvement over baseline

Statistics:
Mean Actual Difficulty: 0.4973
Mean Predicted Difficulty: 0.6284
Std Dev Actual: 0.2201
Std Dev Predicted: 0.1325

**Comment by OmriDan** [comment]
# Classification
### on CS course dataset
model: ModernBERT-base_leetcode_no_rag
Performance Metrics:
Difficulty Mode: categorical
0-0.25; 0.25-0.75; 0.75-1
Accuracy: 0.7042
F1 Macro: 0.3281
F1 Weighted: 0.6299

0-0.33; 0.33-0.67, 0.67-1
Accuracy: 0.5493
F1 Macro: 0.3113
F1 Weighted: 0.4278

model: ModernBERT-large_leetcode_no_rag
Performance Metrics:
0-0.25; 0.25-0.75; 0.75-1
Accuracy: 0.2958
F1 Macro: 0.2127
F1 Weighted: 0.3096

0-0.33; 0.33-0.67, 0.67-1
Accuracy: 0.3380
F1 Macro: 0.2841
F1 Weighted: 0.2998
